File: app/routers/auction.py
from fastapi import APIRouter

from util.tables import Tables
from util.delete import delete
from util.update import update
from util.retrieve import retrieve
from util.insert import insert

from models import AuctionModel


router = APIRouter(prefix="/auctions", tags=['auctions'])

@router.get("/{auction_id}")
def get_auction(
    auction_id: int
):
    success, _, message, items = retrieve(
        table=Tables.Collector.value,
        single=True,
        auction_id=auction_id
    )

    return {"data": items[0], "success": success, "message": message}

@router.get("/")
def get_auctions(
    start_time: str | None = None,
    end_time: str | None = None,
    gt__start_time: str | None = None,
    gt__end_time: str | None = None,
    lt__start_time: str | None = None,
    lt__end_time: str | None = None,
    active: bool | None = None,
    art_id: int | None = None,
):
    success, count, message, items = retrieve(
        table=Tables.Auction.value,
        single=False,
        start_time=start_time,
        end_time=end_time,
        gt__end_time=gt__end_time,
        gt__start_time=gt__start_time,
        lt__end_time=lt__end_time,
        lt__start_time=lt__start_time,
        active=active,
        art_id=art_id
    )

    return {"data": items, "success": success, "message": message, "count": count}


@router.post("/")
def create_new_auction(request_data: AuctionModel):
    success, message = insert(request_data)
    return {"message": message, "success": success}

@router.delete("/{auction_id}")
def delete_auction(auction_id: int):
    success, message = delete(
        table=Tables.Auction.value,
        auction_id=auction_id
    )
    return {"message": message, "success": success}


@router.put("/{auction_id}")
def update_auction(auction_id: int, request_data: AuctionModel):
    success, message = update(
        table=Tables.Auction.value,
        model=request_data,
        auction_id=auction_id
    )
    return {"message": message, "success": success}

File: frontend/api/auction.ts
import {get, post, put, deleteItem, toQueryString} from "./crude";
import type {Auction, ApiReuslt} from "./api_types";

export const getAuction = async (auctionId: number): Promise<ApiReuslt<Auction>> => {
    return get<Auction>(`http://localhost:8000/auctions/${auctionId}`);
};

type AuctionQueryParams = {
    start_time?: string;
    end_time?: string;
    gt__startTime?: string;
    gt__endTime?: string;
    lt__startTime?: string;
    lt__endTime?: string;
    active?: boolean;
    art_id?: number;
};

export const getAuctions = async (params: AuctionQueryParams): Promise<ApiReuslt<Auction[]>> => {
    const queryString = toQueryString(params);
    return get<Auction[]>(`http://localhost:8000/auctions?${queryString}`);
};


export type NewAuctionData = {
    start_time: string;
    end_time: string;
    active: boolean;
    art_id: number;
};

export const createNewAuction = async (data: NewAuctionData): Promise<ApiReuslt<Auction>> => {
    return post<NewAuctionData, Auction>(`http://localhost:8000/auctions`, data);
};


type UpdateAuctionData = {
    start_time?: string;
    end_time?: string;
    active?: boolean;
};

export const updateAuction = async (auctionId: number, data: UpdateAuctionData): Promise<ApiReuslt<Auction>> => {
    return put<UpdateAuctionData, Auction>(`http://localhost:8000/auctions/${auctionId}`, data);
};

export const deleteAuction = async (auctionId: number): Promise<ApiReuslt<null>> => {
    return await deleteItem(`http://localhost:8000/auctions/${auctionId}`, null);
};



